proto_parser/demo/demo_main.cc

```diff
@@ -0,0 +1,401 @@
+  //             field_exists: "PlayExtension.store.click"
+  //             # Condition 2
+  //             field_value {
+  //               path: "PlayExtension.log_source"
+  //               equals_int: 65  # WEB_STORE
+  //             }
+  //             # Condition 3
+  //             ui_element {
+  //               type: STORE_UI_ELEMENT
+  //               and {
+  //                 # Condition 3.1
+  //                 leaf_ui {
+  //                   path:  "type"
+  //                   equals_int: 7305  # AUTHENTICATION_OPT_OUT_CHECKED
+  //                 }
+  //                 # Condition 3.2
+  //                 index_ui {
+  //                   index: -2
+  //                   path: "type"
+  //                   equals_int: 1  # SOMETHING
+  //                 }
+  //               }
+  //             }
+  //           }
+  std::unique_ptr<ProtoValue> ConvertEventPredicateFromFile(
+      absl::string_view path) {
+    ProtoParser parser(ReadTextProtoFromStream(path));
+    EventPredicate event_predicate_prototype;
+    std::unique_ptr<ProtoValue> event_predicate = 
+        parser.Create(event_predicate_prototype);
+    MessageValue* event_predicate_ptr = 
+        static_cast<MessageValue*>(event_predicate.get());  
+    // We are converting event_matcher/ui_element_path_predicate to and here
+    // event_matcher {} => and {}
+    // ui_element_path_predicate{} => and{}
+    std::unique_ptr<ProtoValue> and_field = 
+        absl::make_unique<MessageValue>("and", 0, 0);
+    MessageValue* and_field_ptr = static_cast<MessageValue*>(and_field.get());
+
+    for (const std::unique_ptr<ProtoValue>& field : 
+        event_predicate_ptr->GetFields()) {
+      if (field->GetName() == "event_matcher") {
+        absl::flat_hash_map<std::string, ProtoValue*> field_name_to_value_ptr = 
+            GetFieldNameToValuePtrMap(*static_cast<MessageValue*>(field.get()));
+        if (field_name_to_value_ptr.find("int_comparator") != 
+            field_name_to_value_ptr.end()) {
+          and_field_ptr->AddField(EventMatcherToFieldValue(*field));
+        } else {
+          and_field_ptr->AddField(EventMatcherToFieldExists(*field));
+        }
+        
+      } else if (field->GetName() == "ui_element_path_predicate") {
+        and_field_ptr->AddField(UiElementPathPredicateToUiElement(*field));
+      }
+    }
+
+    return and_field;
+  }
+
+ private:
+
+  // Reads text proto into a string from a file.
+  std::string ReadTextProtoFromStream(absl::string_view path) {
+    std::ifstream ifs;
+    std::string text_proto;
+
+    ifs.open(std::string(path));
+
+    assert(ifs.good());
+
+    while (ifs.good()) {
+      std::string tmp;
+      std::getline(ifs, tmp);
+      text_proto += tmp + "\n";
+    }
+
+    return text_proto;
+  }
+
+  // Copies the comments of an original proto value to a copy, if they're both
+  // primitive types also copy the value.
+  void CopyVal(ProtoValue* copy, const ProtoValue& original) {
```

Comment:
CopyForPrimitiveValue()

Reply:
This also copies for message value though